### PR TITLE
search jobs: allow repo: predicates

### PR DESCRIPTION
**Pull request by stefanhengl**
We allow repo predicates for Search Jobs by removing the corresponding validation checks. At the same time we improve the checks against file predicates.

We now support the following **repo** predicates:
- .has.meta
- .has.file
- .has.content
- .has.path
- .has.commit.after
- .has.description

Notes:
- Why don't we allow file predicates? `file:has.content()` is translated to an AND query in our (query parsing) job layer. In Search Jobs we don't call this code path. A quick copy&paste hack let to Search Jobs breaking in unexpected ways. Hence I want to leave file predicates to a separate investigation and PR.
- The validation in the client and backend are not 100% identical which is a trade-off we accepted for the EAP. The plan is to remove the client validation soon and instead call out to the backend for validation. 

Test plan:
- updated unit tests for client and backend
- manual testing
  - I ran searches with various repo predicates and compared match counts of search jobs with the match counts we display in the Search UI. Repo predicates worked without any issue. This is to be expected because repo predicates are translated to `RepoListOptions` which fits in nicely with our design of Search Jobs.

## Preview 🤩
[Preview Link](https://docs.sourcegraph.com/@sh/search-jobs-enable-repo-predicates)

### CODE REVIEW

**Comment by stefanhengl** [comment]
cc @MaedahBatool tagging you for the 1-line doc change.

File: internal/search/job/jobutil/exhaustive_job.go [hunk 1]
```diff
@@ -46,15 +44,12 @@ func NewExhaustive(inputs *search.Inputs) (Exhaustive, error) {
 		return Exhaustive{}, errors.Errorf("expected a simple expression (no and/or/etc). Got %v", b.Pattern)
 	}
 
-	// no predicates
-	if inputs.Query.Exists(query.FieldRepoHasFile) {
-		return Exhaustive{}, errors.Errorf("repoHasFile is not supported")
```
Comment (keegancsmith) [inline internal/search/job/jobutil/exhaustive_job.go:51]: so we support this predicate now?

File: internal/search/job/jobutil/exhaustive_job_test.go [hunk 2]
```diff
@@ -157,10 +157,11 @@ func TestNewExhaustive_negative(t *testing.T) {
 		// catch-all regex
 		{query: `type:file index:no r:.* .*`, isPatterntypeRegex: true},
 		{query: `type:file index:no r:repo .*`, isPatterntypeRegex: true},
-		// predicates
-		{query: `type:file index:no repohasfile:foo.bar content`},
-		{query: `type:file index:no file:has.content("content")`},
-		{query: `type:file index:no repo:has.path("src") content`},
```
Comment (keegancsmith) [inline internal/search/job/jobutil/exhaustive_job_test.go:163]: should you add some tests for things we do support now?

**Review by keegancsmith — APPROVED** [review]

**Comment by keegancsmith** [comment]
backport?

File: internal/search/job/jobutil/exhaustive_job.go (hunk 1 above)
Comment (stefanhengl) [reply internal/search/job/jobutil/exhaustive_job.go:51]: yes, we support `repo:has.file(...)`

**Pull request merged**